jsgen: fix multiple `var` argument and `addr` issues

zerbina opened the pull request:
## Summary
- fix `addr x` and `p_var(x)` causing an internal error when `x` is a
  for-var or an lvalue-conversion expression (e.g. `int(val)`)
- fix incorrect JS code being generated for `addr (stmt; x)` and
  `p_var( (stmt; x) )`
- fix incorrect JS code being generated for `x.y` where `x` is an immutable
  parameter that is used somwhere as an argument to a `var` parameter
  or as the operand of an `addr`

## Details
- refactor, simplify, and document `genAddr`
- add an extensive test for `var` arguments
- extend the `taddr.nim` test

---

Depends on PR #502 and blocks #498 .

saem commented on compiler/backend/jsgen.nim:
```diff
@@ -1377,53 +1370,47 @@ proc genSymAddr(p: PProc, n: PNode, typ: PType, r: var TCompRes) =
+    genCheckedFieldOp(p, n, takeAddr=true, r)
+  of nkDotExpr:
+    genFieldAddr(p, n, r)
+  of nkBracketExpr:
```
Thinking or loud: This bit really highlights how the AST is just off, it really should be lowered/specialized further. The type based dispatch within the branch seems off.

saem commented on compiler/backend/jsgen.nim:
```diff
@@ -2620,18 +2607,23 @@ proc gen(p: PProc, n: PNode, r: var TCompRes) =
   of nkAddr, nkHiddenAddr:
-    if n.typ.kind in {tyLent}:
+    if mapType(n.typ) != etyBaseIndex:
+      # the operation doesn't produce an addres-like value (e.g. because the
```
```suggestion
      # the operation doesn't produce an address-like value (e.g. because the
```

saem commented on tests/lang_types/var/tvar_arguments.nim:
```diff
@@ -0,0 +1,200 @@
+discard """
+  target: "c js vm"
```
IIRC the lang category runs on all targets by default.

zerbina replied:
Yep, the `lang` category does, but the `lang_xxx` ones do not.

saem commented on tests/lang_types/var/tvar_arguments.nim:
```diff
@@ -0,0 +1,200 @@
+  target: "c js vm"
+  labels: "var_arg"
+  description: '''
+    Tests to make sure arguments are properly passed to var parameter and that
```
These tests are freaking awesome.

saem reviewed (commented):
A+, would try merging.

zerbina committed:
jsgen: fix multiple `var` argument and `addr` issues

## Summary
- fix `addr x` and `p_var(x)` causing an internal error when `x` is a
  for-var or an lvalue-conversion expression (e.g. `int(val)`)
- fix incorrect JS code being generated for `addr (stmt; x)` and
  `p_var( (stmt; x) )`
- fix incorrect JS code being generated for `x.y` where `x` is an immutable
  parameter that is used somwhere as an argument to a `var` parameter
  or as the operand of an `addr`

## Details
- refactor, simplify, and document `genAddr`
- add an extensive test for `var` arguments
- extend the `taddr.nim` test

zerbina commented:
bors r+

zerbina commented:
bors r+
